python/ray/data/expressions.py

```diff
@@ -1042,6 +1073,37 @@ def download(uri_column_name: str) -> DownloadExpr:
     return DownloadExpr(uri_column_name=uri_column_name)
 
 
+@PublicAPI(stability="alpha")
+def monotonically_increasing_id() -> MonotonicallyIncreasingIdExpr:
+    """
+    Create an expression that generates monotonically increasing IDs.
+
+    The generated IDs are guaranteed to be monotonically increasing and unique,
+    but not consecutive. The current implementation puts the task ID in the upper
+    31 bits, and the record number within each task in the lower 33 bits. Records
+    within the block(s) assigned to a task receive consecutive IDs. Note that IDs
+    are not globally ordered across tasks.
+
+    The assumption is that the dataset schedules less than 1 billion tasks, and
+    each task processes less than 8 billion records.
+
+    The function is non-deterministic because its result depends on task IDs.
```

Comment:
Task has to be deterministic assuming its inputs are fixed (and this method is deterministic in that def)